pkg/controllers/provisioning/scheduling/topologygroup.go

```diff
@@ -224,17 +243,34 @@ func (t *TopologyGroup) domainMinCount(domains *scheduling.Requirement) int32 {
 	return min
 }
 
+// nolint:gocyclo
 func (t *TopologyGroup) nextDomainAffinity(pod *v1.Pod, podDomains *scheduling.Requirement, nodeDomains *scheduling.Requirement) *scheduling.Requirement {
 	options := scheduling.NewRequirement(podDomains.Key, v1.NodeSelectorOpDoesNotExist)
-	for domain := range t.domains {
-		if podDomains.Has(domain) && t.domains[domain] > 0 {
-			options.Insert(domain)
+
+	// If we are explicitly selecting on specific node domains ("In" requirement) and the number of node domains
+	// is less than our total domains, this is going to be more efficient to iterate through
+	// This is particularly useful when considering the hostname topology key that can have a
+	// lot of t.domains but only a single nodeDomain
+	if nodeDomains.Operator() == v1.NodeSelectorOpIn && nodeDomains.Len() < len(t.domains) {
+		for _, domain := range nodeDomains.Values() {
+			if count, ok := t.domains[domain]; podDomains.Has(domain) && ok && count > 0 {
+				options.Insert(domain)
+			}
+		}
+	} else {
+		for domain := range t.domains {
+			if podDomains.Has(domain) && t.domains[domain] > 0 {
+				options.Insert(domain)
+			}
 		}
 	}
+	if options.Len() != 0 {
+		return options
+	}
 
-	// If pod is self selecting and no pod has been scheduled yet, we can pick a domain at random to bootstrap scheduling
-
-	if options.Len() == 0 && t.selects(pod) {
+	// If pod is self-selecting and no pod has been scheduled yet, we can pick a domain at random to bootstrap scheduling
+	// This also requires that there not already be some other domain where this a pod in this topology has scheduled
+	if t.selects(pod) && len(t.domains) == len(t.emptyDomains) {
```

Comment:
Just making sure I understand this change: previously, we may have chosen to "bootstrap" a node for a pod with self-affinity if there were no compatible domains with the pod for the affinity term. However, there may have been a pod scheduled in the topology group, and it's domain was just incompatible with this pod's requirements. In this case, previously Karpenter would have created a new node in a compatible domain, but with this change Karpenter would fail to schedule the pod. Are we sure this change matches `kube-scheduler's` behavior? I thought a pod w/ affinty could schedule to a separate domain if the only domain in it's topology group with a matching pod isn't compatible.

Reply:
Discussed offline, there are two actual issues:

We need to iterate over all domains in the topology group, not just node domains. If there is a pod in the topology group in a domain which is compatible with this pod's requirements, we should consider that an option. This is currently missed by only iterating over node domains, since it may be in a domain incompatible with the node. 

Additionally, if there are no options available, we don't want to necessarily short-circuit just because `len(t.domains) == len(t.emptyDomains)`. If there is a pod in the topology group that is **not** compatible with the pod's requirements, it should not be considered.

Reply:
Fixed this by validating if there is a compatible domain that has a pod -- this would prevent us from generating a value for this node and scheduling the pod to the current node.